fl4health/strategies/feddg_ga_with_adaptive_constraint.py

```diff
@@ -0,0 +1,268 @@
+                aggregated train loss before reducing the value of mu. Only applicable if adaptivity is on.
+                Defaults to 5.
+            weighted_train_losses (bool, optional): Determines whether the training losses from the clients should be
+                aggregated using a weighted or unweighted average. These aggregated losses are used to adjust the
+                proximal weight in the adaptive setting. Defaults to False.
+            fairness_metric : FairnessMetric, optional.
+                The metric to evaluate the local model of each client against the global model in order to
+                determine their adjustment weight for aggregation. Can be set to any default metric in
+                FairnessMetricType or set to use a custom metric. Optional, default is
+                FairnessMetric(FairnessMetricType.LOSS).
+            weight_step_size : float
+                The step size to determine the magnitude of change for the adjustment weight. It has to be
+                0 < weight_step_size < 1. Optional, default is 0.2.
+        """
+        if fraction_fit != 1.0 or fraction_evaluate != 1.0:
+            log(
+                WARNING,
+                "fraction_fit or fraction_evaluate are not 1.0. The behaviour of FedDG-GA is unknown in those cases.",
+            )
+
+        self.loss_weight = initial_loss_weight
+        self.adapt_loss_weight = adapt_loss_weight
+
+        if self.adapt_loss_weight:
+            self.loss_weight_delta = loss_weight_delta
+            self.loss_weight_patience = loss_weight_patience
+            self.loss_weight_patience_counter: int = 0
+
+        self.previous_loss = float("inf")
+
+        self.server_model_weights = parameters_to_ndarrays(initial_parameters)
+        initial_parameters.tensors.extend(ndarrays_to_parameters([np.array(initial_loss_weight)]).tensors)
+
+        super().__init__(
+            fraction_fit=fraction_fit,
+            fraction_evaluate=fraction_evaluate,
+            min_fit_clients=min_fit_clients,
+            min_evaluate_clients=min_evaluate_clients,
+            min_available_clients=min_available_clients,
+            evaluate_fn=evaluate_fn,
+            on_fit_config_fn=on_fit_config_fn,
+            on_evaluate_config_fn=on_evaluate_config_fn,
+            accept_failures=accept_failures,
+            initial_parameters=initial_parameters,
+            fit_metrics_aggregation_fn=fit_metrics_aggregation_fn,
+            evaluate_metrics_aggregation_fn=evaluate_metrics_aggregation_fn,
+        )
+
+        if fairness_metric is None:
+            self.fairness_metric = FairnessMetric(FairnessMetricType.LOSS)
+        else:
+            self.fairness_metric = fairness_metric
+
+        self.weight_step_size = weight_step_size
+        assert 0 < self.weight_step_size < 1, f"weight_step_size has to be between 0 and 1 ({self.weight_step_size})"
```

Comment:
nit: should we also do assertions on self.initial_loss_weight, does it also have to be between 0 and 1?

Reply:
So this is actually un-related to the penalty weight parameter, which can be any value we want it to be technically. This is for the way in which the generalization adjustment weights are modified, which is confusing. I'll rename this value to specify that.